tools/tpm2_duplicate.c

```diff
@@ -0,0 +1,423 @@
+    bool result = true;
+    switch (alg) {
+
+    case TPM2_ALG_AES :
+        obj->algorithm = TPM2_ALG_AES;
+        obj->keyBits.aes = 128;
+        obj->mode.aes = TPM2_ALG_CFB;
+        break;
+
+    case TPM2_ALG_NULL :
+        obj->algorithm = TPM2_ALG_NULL;
+        break;
+
+    default:
+        LOG_ERR("The algorithm type input(%4.4x) is not supported!", alg);
+        result = false;
+        break;
+    }
+
+    return result;
+}
+
+static bool load_sym_key(TPM2B_DATA * in_key) {
+    FILE *f;
+    unsigned long file_size = 0;
+    bool result;
+
+    f = fopen(ctx.sym_key_in, "r");
+    if (!f) {
+        LOG_ERR("Could not open file \"%s\", error: %s",
+                ctx.sym_key_in, strerror(errno));
+        return false;
+    }
+    /*
+    * Get the file size and validate that it is the proper AES keysize.
+    */
+    result = files_get_file_size(f, &file_size, ctx.sym_key_in);
+    if (!result) {
+        goto fail;
+    }
+
+    if(file_size != 16) {
+        LOG_ERR("Invalid AES key size, got %lu bytes, expected 16",
+                file_size);
+        result = false;
+    }
+    else
+    {
+        in_key->size = file_size;
+        result = files_read_bytes(f, in_key->buffer, in_key->size);
+    }
+
+fail:
+    fclose(f);
+    return result;
+}
+
+int tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
+    UNUSED(flags);
+
+    int rc = 1;
+    bool result;
+    TPMT_SYM_DEF_OBJECT sym_alg;
+    TPM2B_DATA in_key;
+    TPM2B_DATA* out_key;
+    TPM2B_PRIVATE* duplicate;
+    TPM2B_ENCRYPTED_SECRET* outSymSeed;
+
+    result = check_options();
+    if (!result) {
+        goto out;
+    }
+
+    result = load_object(ectx, ctx.object_arg, &ctx.object_context);
+    if(!result) {
+        goto out;
+    }
+
+    result = load_object(ectx, ctx.new_parent_object_arg, &ctx.new_parent_object_context);
+    if(!result) {
+        goto out;
+    }
+
+    if (ctx.flags.p) {
+        result = tpm2_auth_util_from_optarg(ectx, ctx.object_auth_str,
+            &ctx.auth.session_data, &ctx.auth.session);
+        if (!result) {
+            LOG_ERR("Invalid authorization, got\"%s\"", ctx.object_auth_str);
+            goto out;
+        }
+    }
+
+    result = set_key_algorithm(ctx.key_type, &sym_alg);
+    if(!result) {
+        goto out;
+    }
+
+    if(ctx.flags.i) {
+        result = load_sym_key(&in_key);
+        if(!result) {
+            goto out;
+        }
+    }
+
+    out_key = NULL;
+
+    result = do_duplicate(ectx,
+        ctx.flags.i ? &in_key : NULL,
+        &sym_alg,
+        ctx.flags.o ? &out_key : NULL,
+        &duplicate,
+        &outSymSeed);
+    if (!result) {
+        goto out;
+    }
+
+
+    result = tpm2_session_save(ectx, ctx.auth.session, NULL);
+    if (!result) {
+        goto free_out;
+    }
+
+    /* Maybe a false positive from scan-build but we'll check out_key anyway */
+    if (ctx.flags.o) {
+        if(out_key == NULL) {
```

Comment:
Yeah scan-build can be a pain. #include <assert.h> and assert(out_key) if you get these warnings. Does it go away when you declare out_key = NULL rather than setting it on L361.